resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ExecutorRunnable.scala

```diff
@@ -246,13 +253,57 @@ private[yarn] class ExecutorRunnable(
       sys.env.get("SPARK_USER").foreach { user =>
         val containerId = ConverterUtils.toString(c.getId)
         val address = c.getNodeHttpAddress
-        val baseUrl = s"$httpScheme$address/node/containerlogs/$containerId/$user"
 
-        env("SPARK_LOG_URL_STDERR") = s"$baseUrl/stderr?start=-4096"
-        env("SPARK_LOG_URL_STDOUT") = s"$baseUrl/stdout?start=-4096"
+        val customLogUrl = sparkConf.get(config.CUSTOM_LOG_URL)
+
+        val envNameToFileNameMap = Map("SPARK_LOG_URL_STDERR" -> "stderr",
+          "SPARK_LOG_URL_STDOUT" -> "stdout")
+        val logUrls = ExecutorRunnable.buildLogUrls(customLogUrl, httpScheme, address,
+          clusterId, containerId, user, envNameToFileNameMap)
+        logUrls.foreach { case (envName, url) =>
+          env(envName) = url
+        }
       }
     }
 
     env
   }
 }
+
+private[yarn] object ExecutorRunnable {
+  def buildLogUrls(
+      logUrlPattern: String,
+      httpScheme: String,
+      nodeHttpAddress: String,
+      clusterId: Option[String],
+      containerId: String,
+      user: String,
+      envNameToFileNameMap: Map[String, String]): Map[String, String] = {
+    val optionalPathVariable: Map[String, Option[String]] = Map("{{ClusterId}}" -> clusterId)
+    val pathVariables: Map[String, String] = Map("{{HttpScheme}}" -> httpScheme,
```

Comment:
I must be missing something why are some optional and others not, if a user want to do a custom url couldn't they not have containerid or user for instance?

Reply:
My intention was to separate parameters by `mandatory` and `optional`. Mandatory parameters should be available for all the cases, whereas optional parameters are available only when Hadoop/YARN has configured like that.